[BEP036] Tabular phenotype example datasets

Arshitha opened the pull request:
Added `pheno001` and `pheno002` example dataset inspired by [ds004215](https://openneuro.org/datasets/ds004215/versions/2.0.1) on OpenNeuro but significantly modified to keep it simple and easy to convey the various use cases proposed in [BEP036](https://bids.neuroimaging.io/bep036). 

Use cases covered (and to be added to this PR):
- [x] `pheno001` - Single session with both phenotype and imaging data
- [x] `pheno002` - Two sessions with one imaging data only session
- [x] `pheno003` - Two sessions with one phenotype data only session
- [x] `pheno004` - Two _sets_ of sessions. One set of sessions (e.g. screening, baseline, followup, etc) for phenotype data and another set of sessions (e.g. 01, 02, etc) for imaging data. 

Still in draft state but would appreciate any and all feedback. 

Pinging co-contributors: @ericearl @SamGuay @surchs 

- BIDS Specification PR: https://github.com/bids-standard/bids-specification/pull/2123
- BIDS Validator PR: https://github.com/bids-standard/bids-validator/pull/366

Arshitha committed:
adding 2 example datasets for bep036

ericearl reviewed (dismissed):
This is great, thanks! I'm guessing you left it in draft state becausew of pheno001 and pheno002, right?

I think we should remove the `age_at_visit` column/field from all `phenotype/` measurement tools and instead provide a root-level sessions file with that field. Should we maybe take that a step farther and RECOMMEND or say it's OPTIONAL to add `age` to the sessions file?

Arshitha commented:
I like that idea. It's redundant information that can be aggregated to sessions level, and can be a recommendation in the BEP.

Arshitha commented:
It's in Draft state because I haven't prepared `pheno003` and `pheno004` but yes, all four example datasets will violate the contribution guidelines.

Arshitha committed:
adding a dataset ; minor changes to existing datasets

ericearl commented on pheno001/phenotype/ace.tsv:
Does this one need the `session_id` column to retain the second row?

Arshitha replied:
This is an example of  "all participants have one session across modalities" case.  This does not require a `session_id` or sessions file if I'm understanding the proposal correctly.

ericearl replied:
Oh, so the original second row was a mistake. Got it! Carry on and ignore this comment.

Arshitha replied:
yep!

ericearl commented on pheno001/participants.json:
```diff
@@ -0,0 +1,37 @@
+      "f": "Female"
+    }
+  },
+  "age_at_visit": {
```
```suggestion
  "age": {
```

Arshitha replied:
In this case, `age_at_visit` would be the same as `age` and that's the reason I changed it to `age_at_visit`.

ericearl replied:
I think it's still fine to keep as `age`, but I would let @SamGuay and/or @surchs be the tie-breakers.

ericearl commented on pheno001/participants.tsv:
```diff
@@ -0,0 +1,3 @@
+participant_id	sex	age_at_visit	MRI	ace	demographics
```
```suggestion
participant_id	sex	age	MRI	ace	demographics
```

Arshitha replied:
same as previous comment

ericearl replied:
cleaning this up for a soon merge.

ericearl commented on pheno002/sessions.json:
```diff
@@ -0,0 +1,19 @@
+      "02": "Second imaging only visit."
+    }
+  },
+  "age_at_visit": {
```
```suggestion
  "age": {
```

ericearl commented on pheno002/sessions.tsv:
```diff
@@ -0,0 +1,5 @@
+participant_id	session_id	age_at_visit	acq_time
```
```suggestion
participant_id	session_id	age	acq_time
```

ericearl commented on pheno003/sessions.json:
```diff
@@ -0,0 +1,16 @@
+      "followup": "Second phenotype data collection visit."
+    }
+  },
+  "age_at_visit": {
```
```suggestion
  "age": {
```

ericearl commented on pheno003/sessions.tsv:
```diff
@@ -0,0 +1,4 @@
+participant_id	session_id	age_at_visit
```
```suggestion
participant_id	session_id	age
```

Arshitha committed:
Update pheno002/sessions.tsv

Co-authored-by: Eric Earl <eric.earl@nih.gov>

Arshitha committed:
Update pheno003/sessions.json

Co-authored-by: Eric Earl <eric.earl@nih.gov>

Arshitha committed:
Update pheno002/sessions.json

Co-authored-by: Eric Earl <eric.earl@nih.gov>

Arshitha committed:
Update pheno003/sessions.tsv

Co-authored-by: Eric Earl <eric.earl@nih.gov>

ericearl committed:
Update pheno 001-004 participants.tsv files and add pheno004

- All participants.tsv files have been simplified.
- pheno004 has become instead an example of some imaging-only, some phenotype-only, and some with both data

ericearl commented:
I hijacked the not yet created `pheno004` for use with some current bids-validator testing needs at https://github.com/bids-standard/bids-specification/pull/2044. Can we make this a non-draft PR and get it merged hopefully? @effigies @Arshitha @SamGuay @surchs

effigies commented:
Please set the `BIDS_SCHEMA` environment variable to https://bids-specification--2044.org.readthedocs.build/en/2044/schema.json here:

https://github.com/bids-standard/bids-examples/blob/e52f77fcd1abe7f234ecaec3542082279b043f8a/.github/workflows/validate_datasets.yml#L114

Please also add pheno004 to be skipped on legacy and stable:

https://github.com/bids-standard/bids-examples/blob/e52f77fcd1abe7f234ecaec3542082279b043f8a/.github/workflows/validate_datasets.yml#L98-L101

https://github.com/bids-standard/bids-examples/blob/e52f77fcd1abe7f234ecaec3542082279b043f8a/.github/workflows/validate_datasets.yml#L103-L106

ericearl commented:
@effigies Is [that comment just above here](https://github.com/bids-standard/bids-examples/pull/465#issuecomment-2640990645) a note for me? I'm confused by most of it and don't feel safe editing those files as-is. If you need me to take care of that, can I sit with you, Ross, or Nell to figure it out or have it explained to me enough to be able to do the work?

effigies committed:
Merge branch 'master' into master

effigies committed:
chore(ci): Use draft schema, shield pheno004 from older validators

effigies commented:
Okay, I did what I asked. It looks like there are issues in the schema that need to be addressed, but also there are unrelated issues in pheno001-003: https://github.com/bids-standard/bids-examples/actions/runs/13188395001/job/36815880378?pr=465

ericearl commented:
This is super-helpful @effigies, thank you! I'm bringing the errors out of the logs here for us (@Arshitha @SamGuay @surchs):

```
# pheno001

	[ERROR] MISSING_DATASET_DESCRIPTION A dataset_description.json file is required in the root of the dataset
		

	Please visit https://neurostars.org/search?q=MISSING_DATASET_DESCRIPTION for existing conversations about this issue.

	[ERROR] JSON_INVALID Not a valid JSON file.
		/sub-01/anat/sub-01_T1w.json
		/sub-01/anat/sub-01_T1w.json

		2 more files with the same issue

	Please visit https://neurostars.org/search?q=JSON_INVALID for existing conversations about this issue.

# pheno002

	[ERROR] MISSING_DATASET_DESCRIPTION A dataset_description.json file is required in the root of the dataset
		

	Please visit https://neurostars.org/search?q=MISSING_DATASET_DESCRIPTION for existing conversations about this issue.

	[ERROR] TSV_COLUMN_ORDER_INCORRECT Some TSV columns are in the incorrect order
		session_id
		/sessions.tsv - Column 0 (starting from 0) found at index 1.

	Please visit https://neurostars.org/search?q=TSV_COLUMN_ORDER_INCORRECT for existing conversations about this issue.

	[ERROR] TSV_INDEX_VALUE_NOT_UNIQUE An index column(s) was specified for the tsv file and not all of the values for it are unique.
		/sessions.tsv - Row: 4, Value: 01
		/sessions.tsv - Row: 5, Value: 02

	Please visit https://neurostars.org/search?q=TSV_INDEX_VALUE_NOT_UNIQUE for existing conversations about this issue.

	[ERROR] JSON_INVALID Not a valid JSON file.
		/sub-01/ses-01/anat/sub-01_ses-01_T1w.json
		/sub-01/ses-01/anat/sub-01_ses-01_T1w.json

		6 more files with the same issue

	Please visit https://neurostars.org/search?q=JSON_INVALID for existing conversations about this issue.

# pheno003

	[ERROR] MISSING_DATASET_DESCRIPTION A dataset_description.json file is required in the root of the dataset
		

	Please visit https://neurostars.org/search?q=MISSING_DATASET_DESCRIPTION for existing conversations about this issue.

	[ERROR] TSV_COLUMN_ORDER_INCORRECT Some TSV columns are in the incorrect order
		session_id
		/sessions.tsv - Column 0 (starting from 0) found at index 1.

	Please visit https://neurostars.org/search?q=TSV_COLUMN_ORDER_INCORRECT for existing conversations about this issue.

	[ERROR] TSV_INDEX_VALUE_NOT_UNIQUE An index column(s) was specified for the tsv file and not all of the values for it are unique.
		/sessions.tsv - Row: 4, Value: baseline

	Please visit https://neurostars.org/search?q=TSV_INDEX_VALUE_NOT_UNIQUE for existing conversations about this issue.

	[ERROR] JSON_INVALID Not a valid JSON file.
		/sub-01/ses-baseline/anat/sub-01_ses-baseline_T1w.json
		/sub-01/ses-baseline/anat/sub-01_ses-baseline_T1w.json

		2 more files with the same issue

	Please visit https://neurostars.org/search?q=JSON_INVALID for existing conversations about this issue.

```

Arshitha committed:
fixing some bids validation errors

Arshitha commented:
@ericearl - For `pheno004`, I added `dataset_description.json` and replaced empty nifti and json with non-empty ones from an openneuro dataset. Lmk if you'd like me to undo those changes.

Arshitha committed:
adding pheno005 as multi-run example

Arshitha committed:
adding run info to sessions.tsv

Arshitha committed:
minor change

ericearl commented on pheno005/dataset_description.json:
```diff
@@ -0,0 +1,12 @@
+{
+	"Name": "BEP036 example dataset 05: Two multi-run sessions with one imaging-only session.",
+	"BIDSVersion": "1.9.0",
```
Should this be the latest BIDSVersion, 1.10.0?

Arshitha replied:
changed!

ericearl commented on pheno005/phenotype/demographics.json:
```diff
@@ -0,0 +1,75 @@
+			"followup": "Second phenotype data collection visit."
+		}
+	},
+	"run_id": {
```
I wonder whether it should be `run_id`, `run_index`, or just `run`? See: https://bids-specification.readthedocs.io/en/stable/glossary.html#index-common_principles. I would be inclined to have it as just `run`.

Arshitha replied:
I don't have strong feelings so I'll go with `run` on this one.

ericearl commented on pheno005/phenotype/demographics.json:
```diff
@@ -0,0 +1,75 @@
+	"run_id": {
+		"Description": "ID assigned to a run for a given session.",
+		"Levels": {
+			"01": "First run of the session.",
+			"02": "Second run of the session."
```
These should be `run-01` and `run-02`, respectively. The same should be in the TSV.

ericearl commented on pheno005/phenotype/demographics.json:
```diff
@@ -0,0 +1,75 @@
+		"Description": "Age of the participant.",
+		"Units": "years"
+	},
+	"sex": {
```
I'm conflicted on whether sex at birth should go in the participants.tsv or the demographics.tsv. I know OpenNeuro crawls participants.tsv files better right now to improve its search functionality, which may be justification enough to move this. Anyway, this is a point of discussion and I am open to other ideas.

Arshitha replied:
@ericearl Based on our slack discussion, I think we can leave the demographics info in the demographics file with participants.tsv being a list of unique participant IDs?

ericearl commented on pheno005/sessions.json:
```diff
@@ -0,0 +1,12 @@
+	"session_id": {
+		"Description": "ID assigned to participant's visit.",
+		"Levels": {
+			"baseline": "First imaging and phenotype data collection visit.",
+			"followup": "Second phenotype data collection visit."
```
These should be `ses-baseline` and `ses-followup`, respectively. The same should be in the TSV.

ericearl commented on pheno005/phenotype/ace.json:
```diff
@@ -0,0 +1,99 @@
+	"session_id": {
+		"Description": "ID assigned to participant's visit.",
+		"Levels": {
+			"baseline": "First imaging and phenotype data collection visit.",
+			"followup": "Second phenotype data collection visit."
```
These should be ses-baseline and ses-followup, respectively. The same should be in the TSV.

ericearl commented on pheno005/phenotype/ace.json:
```diff
@@ -0,0 +1,99 @@
+	"run_id": {
+		"Description": "ID assigned to a run for a given session.",
+		"Levels": {
+			"01": "First run of the session.",
+			"02": "Second run of the session."
```
These should be `run-01` and `run-02`, respectively. The same should be in the TSV.

Arshitha commented on pheno005/sub-01/ses-baseline/anat/sub-01_ses-baseline_T1w.nii.gz:
@ericearl The validator was complaining about this so I replaced empty files with actual Niftis (from openneuro) in the other examples and this new example.

Arshitha committed:
Update dataset_description.json

Arshitha committed:
Update dataset_description.json

Arshitha committed:
Update dataset_description.json

Arshitha committed:
Update dataset_description.json

Arshitha committed:
Update ace.tsv

Arshitha committed:
minor changes

Arshitha committed:
resolving merge conflicts

Arshitha commented on pheno005/sub-02/ses-baseline/anat/sub-02_ses-baseline_T1w.nii.gz:
See the above comment.

Arshitha committed:
changes based on code review

Arshitha commented:
@christinerogers `pheno005` is now available for review as the multi-run representative example!

effigies committed:
Merge remote-tracking branch 'upstream/master' into Arshitha/master

Arshitha committed:
Modify Pheno001
- Remove demographics.{tsv,json}
- Add demographics information to participants.tsv
- Enrich session.{tsv,json} to include age at visit which was previously in the demographics files

Arshitha committed:
Modify Pheno002

Arshitha committed:
Modify Pheno003

Arshitha committed:
Modify Pheno004

Arshitha committed:
Modify Pheno005

Arshitha commented:
@ericearl I've modified the example datasets and split the changes in 5 separate commits for easier review. Following changes were made:

- Remove `demographics.{tsv,json}` files.
- Add age at visit information in `sessions.{tsv,json}` files.
- Add other demographic information like sex, age (at baseline), race, etc in `participants.{tsv,json}` files. 
- Add "AdditionalValidation" option in `dataset_description.json`
- Modify `README.md` with updated file tree

Arshitha committed:
Replace non-empty nifti with zero-bytes nifti

ericearl committed:
Update pheno00* examples

ericearl committed:
Update validate_datasets.yml

Updates to validate pheno datasets with new schema for pheno on  bids-spec PR 2123.

ericearl committed:
Update validate_datasets.yml

Trying to succeed to validate the dev validator BIDS_SCHEMA.

ericearl committed:
Update pheno0005 sessions files

Almost forgot the spec calls for run_id and not just run as the column name.

ericearl committed:
Update pheno examples

Added in ses- to a few ace.tsv/json session_id entries.

Arshitha committed:
Fix tab-separation issue

Arshitha committed:
Merge pull request #1 from ericearl/bep036

Update pheno00* examples

Arshitha committed:
Add pheno006 example. No sessions file example.

Arshitha committed:
Update pheno006/README.md

Co-authored-by: Eric Earl <eric.earl@nih.gov>

effigies committed:
fix: Make AdditionalValidation a list

effigies committed:
fix: Remove trailing spaces

ericearl committed:
pheno examples update to correspond with the (hopefully) final BEP036

Arshitha committed:
Merge pull request #2 from ericearl/patch-1

pheno examples update to correspond with the (hopefully) final BEP036

ericearl commented on pheno006/README.md:
```diff
@@ -0,0 +1,28 @@
+# Pheno006 example dataset: Two sessions with one imaging-only session. No sessions files
+
+This dataset contains imaging and phenotypic data from the baseline session and phenotypic data from followup sessions. This an example of a case where adding sessions files does not provide any new information. In other words, a sessions file is not strictly necessary here. The file tree is as follows:
```
```suggestion
This dataset contains imaging and phenotypic data from the baseline session and phenotypic data from followup sessions. This an example of a case where adding sessions files does not provide any new information. In other words, a sessions file is not strictly necessary here. Note how these sessions files do not contain `acq_time`, which will trigger a warning in the validator. The file tree is as follows:
```

ericearl commented on pheno001/phenotype/tool-ACE_phenotype.json:
```diff
@@ -0,0 +1,85 @@
+{
```
```suggestion
{
  "MeasurementToolMetadata": {
    "Description": "Adverse Childhood Experience Questionnaire for Adults"
  },
```

ericearl commented on pheno002/phenotype/tool-ACE_phenotype.json:
```diff
@@ -0,0 +1,92 @@
+{
```
```suggestion
{
  "MeasurementToolMetadata": {
    "Description": "Adverse Childhood Experience Questionnaire for Adults"
  },
```

ericearl commented on pheno003/phenotype/tool-ACE_phenotype.json:
```diff
@@ -0,0 +1,92 @@
+{
```
```suggestion
{
  "MeasurementToolMetadata": {
    "Description": "Adverse Childhood Experience Questionnaire for Adults"
  },
```

ericearl commented on pheno005/phenotype/tool-ACE_phenotype.json:
```diff
@@ -0,0 +1,99 @@
+{
```
```suggestion
{
  "MeasurementToolMetadata": {
    "Description": "Adverse Childhood Experience Questionnaire for Adults"
  },
```

ericearl commented on pheno006/phenotype/tool-ACE_phenotype.json:
```diff
@@ -0,0 +1,92 @@
+{
```
```suggestion
{
  "MeasurementToolMetadata": {
    "Description": "Adverse Childhood Experience Questionnaire for Adults"
  },
```

ericearl committed:
Apply suggestions from code review

Updates thanks to Ross taking a look at the AdditionalValidation vs the current bids-examples. Thanks Ross!

effigies committed:
Merge remote-tracking branch 'upstream/master' into Arshitha/master

yarikoptic commented on pheno001/phenotype/tool-ACE_phenotype.json:
```diff
@@ -0,0 +1,88 @@
+{
+  "MeasurementToolMetadata": {
+    "Description": "Adverse Childhood Experience Questionnaire for Adults"
+  },
```
are we doing that fun dance again of mixing in description of columns **and** additional metadata?  

All of that could be avoided if we followed somewhat established approach for any `{entity}` to have an `{entity_plural}.tsv` which would have description per each value of that entity. So here if it was e.g. `pheno-ace.{tsv,json}` and then `phenotype.tsv` (at top level of within `phenotype/` folder -- doesn't matter  due to Inheritance principle) with `phenotype_id` and `description` columns -- everything would have been nice and consistent.

I really do not get why there is ongoing brewing of some non-BIDS-like cocktail in BEP036.

Related issues backrefs:
- https://github.com/bids-standard/bids-specification/issues/2283
- https://github.com/bids-standard/bids-2-devel/issues/97 (relating to "seggregated" representations)

ericearl committed:
Update pheno00* example datasets

Updated to satisfy the new IndexColumns requirement for tabular phenotypic and assessment data.

Arshitha committed:
Merge pull request #4 from ericearl/eric-indexcolumns-patch

ericearl committed:
Update pheno00* examples

Many "Demographics" to "demographics" renames and one "age" field transfer.

ericearl committed:
Modify README header for Pheno006 dataset

Updated the README header for clarity.

ericearl committed:
Update pheno00* tool-demographics_phenotype.json files

Updated to include MeasurementToolMetadata.
